// @linthtml/core/dist/src/rules/id-no-dup/index.js
import { is_tag_node, attribute_value, has_non_empty_attribute } from "@linthtml/dom-utils";
const RULE_NAME = "id-no-dup";
function lint(node, _config, { report }) {
    if (is_tag_node(node) && has_non_empty_attribute(node, "id")) {
        const id = attribute_value(node, "id");
        // TODO: Remove after `raw-ignore-text` refacto
        if (/^¤+$/.test(id.chars)) {
            return;
        }
        // node has a duplicate id
        // @ts-expect-error To remove once moved to visitor pattern
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
        const saved_id = this.idMap.get(id.chars);
        if (saved_id) {
            report({
                code: "E012",
                position: id.loc,
                meta: {
                    data: {
                        id: id.chars,
                        line: saved_id.loc.start.line,
                        column: saved_id.loc.start.column
                    }
                }
            });
        }
        // if we haven't seen the id before, remember it
        // and pass the node
        // @ts-expect-error To remove once moved to visitor pattern
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
        this.idMap.set(id.chars, id);
    }
}
function end() {
    // wipe previous table
    // @ts-expect-error To remove once moved to visitor pattern
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    this.idMap = new Map();
    return [];
}
export default {
    name: RULE_NAME,
    lint,
    end,
    idMap: new Map()
};
