// @linthtml/core/dist/src/rules/href-style/index.js
import { is_tag_node, has_attribute, attribute_value, get_attribute } from "@linthtml/dom-utils";
import { create_list_value_validator } from "../../validate_option.js";
const RULE_NAME = "href-style";
function lint(node, format, { report }) {
    var _a;
    if (is_tag_node(node) && node.name === "a") {
        // TODO: Should return an issue, since a without href is bad
        // TODO: Check empty href attributes
        if (has_attribute(node, "href") === false) {
            return;
        }
        const href_value = (_a = attribute_value(node, "href")) === null || _a === void 0 ? void 0 : _a.chars;
        // TODO: Remove after `raw-ignore-text` refacto
        if (/^¤+$/.test(href_value)) {
            return;
        }
        // Allow fragment-only URLs
        const is_fragment = href_value.startsWith("#");
        // Link must be absolute if specified format is absolute
        const is_absolute = href_value.search("://") !== -1;
        const match_format = is_absolute === (format === "absolute");
        if (is_fragment === false && match_format === false) {
            const attribute = get_attribute(node, "href");
            report({
                code: "E009",
                position: attribute.loc,
                meta: {
                    data: {
                        format
                    }
                }
            });
        }
    }
}
export default {
    name: RULE_NAME,
    lint,
    validateConfig: create_list_value_validator(RULE_NAME, ["absolute", "relative"], false)
};
