// @linthtml/dom-utils/dist/index.cjs
'use strict';

var is_void_node = require('./is_void_node.cjs');
var is_labelable = require('./is_labelable.cjs');
var check_lang_attribute = require('./check_lang_attribute.cjs');
var attr_parse = require('./attr_parse.cjs');
var boolean_attrs = require('./boolean_attrs.cjs');
var tags = require('./tags.cjs');
var text_node = require('./text_node.cjs');
var dom_elements = require('./dom_elements.cjs');
require('domelementtype');
require('node:util');
require('domhandler');



exports.is_void_node = is_void_node.is_void_node;
exports.is_labelable = is_labelable.is_labelable;
exports.check_lang_attribute = check_lang_attribute.check_lang_attribute;
exports.parse_HTML_attributes = attr_parse.parse_HTML_attributes;
exports.BOOLEAN_ATTRIBUTES = boolean_attrs.BOOLEAN_ATTRIBUTES;
exports.is_boolean_attribute = boolean_attrs.is_boolean_attribute;
exports.attribute_has_value = tags.attribute_has_value;
exports.attribute_value = tags.attribute_value;
exports.get_attribute = tags.get_attribute;
exports.get_classes = tags.get_classes;
exports.has_attribute = tags.has_attribute;
exports.has_non_empty_attribute = tags.has_non_empty_attribute;
exports.has_parent_node = tags.has_parent_node;
exports.is_comment_node = tags.is_comment_node;
exports.is_directive_node = tags.is_directive_node;
exports.is_self_closing = tags.is_self_closing;
exports.is_tag_node = tags.is_tag_node;
exports.is_text_node = tags.is_text_node;
exports.node_tag_name = tags.node_tag_name;
exports.get_text_lines = text_node.get_text_lines;
exports.is_newline_only = text_node.is_newline_only;
exports.CharValue = dom_elements.CharValue;
exports.Document = dom_elements.Document;
exports.Element = dom_elements.Element;
exports.Node = dom_elements.Node;
exports.NodeAttribute = dom_elements.NodeAttribute;
exports.Position = dom_elements.Position;
exports.Range = dom_elements.Range;
